### PR TITLE
Login Manager

### PR DESCRIPTION
**Pull request by awschristou**
## Description

This is largely a rearrangement of existing code into more manageable/testable/isolateable pieces.

* Improved the Toolkit Login/Logout code flows (LoginManager)
* moved credential creation and MFA prompts to a new file `credentialsCreator.ts` -- this will be replaced with a new Provider system in an upcoming change
* Removed Credentials retrieval from AwsContext (moved to credentialsCreator). AwsContext now only retrieves the currently active Credentials
* Replaced CredentialManager with a "dumb" Credentials Cache (CredentialsStore)
* credentials that do not successfully verify are removed from credentials cache (fixes a few open issues)

## Motivation and Context

Preparation for adding a new credentials provider system; the current credentials related code is hard to follow along with.

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/credentials/credentialsCreator.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as nls from 'vscode-nls'
+const localize = nls.loadMessageBundle()
+
+import * as AWS from 'aws-sdk'
+import * as vscode from 'vscode'
+
+const ERROR_MESSAGE_USER_CANCELLED = localize(
```
Comment (bryceitoc9) [inline src/credentials/credentialsCreator.ts]: Is this a new pattern we want to use going forward? I don't see why this shouldn't just be an inline localize call.

File: src/credentials/loginManager.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    public constructor(private readonly awsContext: AwsContext) {}
+
+    /**
+     * Establishes a Credentials for the Toolkit to use. Essentially the Toolkit becomes "logged in".
+     * If an error occurs while trying to set up and verify these credentials, the Toolkit is "logged out".
+     */
+    public async login(credentialsId: string): Promise<void> {
+        try {
+            const credentials = await this.credentialsStore.getCredentialsOrCreate(credentialsId, createCredentials)
+            if (!credentials) {
```
Comment (bryceitoc9) [inline src/credentials/loginManager.ts]: I don't think this will ever hit in its current state. If a credential isn't found, AWSCredentialProviderChain will throw an error: https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/CredentialProviderChain.html#resolvePromise-property

File: src/extension.ts [hunk 3]
```diff
@@ -49,9 +50,15 @@ export async function activate(context: vscode.ExtensionContext) {
         const awsContextTrees = new AwsContextTreeCollection()
         const resourceFetcher = new DefaultResourceFetcher()
         const regionProvider = new DefaultRegionProvider(context, resourceFetcher)
+        const loginManager = new LoginManager(awsContext)
```
Comment (bryceitoc9) [inline src/extension.ts]: Why not create a LoginManager in DefaultAWSContextCommands and loginWithMostRecentCredentials? Both of those took the awsContext before (DefaultAWSContextCommands still does) and can construct a LoginManager from that.

File: src/credentials/credentialsStore.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+        return credentials
+    }
+
+    /**
+     * Evicts credentials from storage
+     */
+    public invalidateCredentials(credentialsId: string) {
+        // tslint:disable-next-line:no-dynamic-delete
+        delete this.credentialsCache[credentialsId]
```
Comment (bryceitoc9) [inline src/credentials/credentialsStore.ts]: Kind of silly that we need the tslint ignore...is there a specific reason to delete instead of setting to undefined? Was this the cause of the overaggressive cache?

File: src/credentials/activation.ts [hunk 5]
```diff
@@ -26,20 +26,14 @@ export async function initialize(parameters: CredentialsInitializeParameters): P
...
-    const currentProfile = toolkitSettings.readSetting(profileSettingKey, '')
-    if (currentProfile) {
-        // check to see if current user is valid
-        const successfulLogin = await UserCredentialsUtils.addUserDataToContext(currentProfile, awsContext)
-        if (!successfulLogin) {
-            await UserCredentialsUtils.removeUserDataFromContext(awsContext)
-            // tslint:disable-next-line: no-floating-promises
-            UserCredentialsUtils.notifyUserCredentialsAreBad(currentProfile)
-        }
+    const previousCredentialsId = toolkitSettings.readSetting(profileSettingKey, '')
```
[the remark this answers is not in the document]
Comment (awschristou) [reply src/credentials/activation.ts]: Not at this time. I'd like to do a dedicated pass to remove `profile` terminology when the new credentials provider system is implemented.

File: src/credentials/credentialsCreator.ts (hunk 1 above)
Comment (awschristou) [reply src/credentials/credentialsCreator.ts]: Anywhere we have static text (eg: without tokens) it doesn't make sense to re-resolve the text each time. We do this in a few places, but not consistently.

File: src/credentials/credentialsStore.ts (hunk 4 above)
Comment (awschristou) [reply src/credentials/credentialsStore.ts]: Calls to invalidate credentials from a cache is new in this PR, which addresses the overaggressive cache problems indeed.

I leaned towards using delete over setting to undefined, so that the map didn't contain blank entries, but I don't think there is an impact either way. I'm mostly indifferent on this one.

File: src/credentials/loginManager.ts (hunk 2 above)
Comment (awschristou) [reply src/credentials/loginManager.ts]: It shouldn't. This is defensive programming against getCredentialsOrCreate and/or createCredentials ever returning undefined.

File: src/extension.ts (hunk 3 above)
Comment (awschristou) [reply src/extension.ts]: The toolkit should only have one loginManager. The LoginManager contains the credentials cache, so if you instantiate more than one, each instance would have its own credentials cache.

File: src/credentials/loginManager.ts (hunk 2 above)
Comment (bryceitoc9) [reply src/credentials/loginManager.ts]: Perhaps note this in a comment and change the error message? Error message could just be something simple like "No credentials found" that encompasses an undefined return.

Comment (awschristou) [reply src/credentials/loginManager.ts]: The code feels self explanatory here, I'm not seeing the value in rehashing what the code says.

I'll update the error though, because "create" does not properly reflect the operation.

**Comment by awschristou** [comment]
Proceeding with this PR's merge. The integration tests are failing on a non-related SAM dependency, and those tests will not pass until a new version of the SAM CLI is released.
